app/src/main/java/org/onionshare/android/server/WebserverManager.kt

```diff
@@ -0,0 +1,116 @@
+import io.ktor.server.netty.Netty
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import java.security.SecureRandom
+import javax.inject.Inject
+
+internal const val PORT: Int = 17638
+
+class WebserverManager @Inject constructor() {
+
+    enum class State { STARTING, STARTED, STOPPING, STOPPED }
+
+    private val _state = MutableStateFlow(State.STOPPED)
+    val state: StateFlow<State> = _state
+
+    private val secureRandom = SecureRandom()
+    private var server: ApplicationEngine? = null
+
+    fun start(sendPage: SendPage) {
+        _state.value = State.STARTING
+        val staticPath = getStaticPath()
+        server = embeddedServer(Netty, PORT, watchPaths = emptyList()) {
+            install(CallLogging)
+            install(Pebble) {
+                loader(ClasspathLoader().apply { prefix = "assets/templates" })
+            }
+            installStatusPages(staticPath)
+            addListener()
+            routing {
+                defaultRoutes(staticPath)
+                sendRoutes(sendPage, staticPath)
+            }
+        }.also { it.start() }
+    }
+
+    fun stop() {
+        _state.value = State.STOPPING
+        server?.stop(1_000, 2_000)
+    }
+
+    private fun getStaticPath(): String {
+        val staticSuffixBytes = ByteArray(16).apply { secureRandom.nextBytes(this) }
+        val staticSuffix =
+            Base64.encodeToString(staticSuffixBytes, NO_PADDING or URL_SAFE).trimEnd()
+        return "/static_$staticSuffix"
+    }
+
+    private fun Application.addListener() {
+        environment.monitor.subscribe(ApplicationStarted) {
+            _state.value = State.STARTED
+        }
+        environment.monitor.subscribe(ApplicationStopped) {
+            _state.value = State.STOPPED
+        }
+    }
+
+    private fun Application.installStatusPages(staticPath: String) {
+        install(StatusPages) {
+            status(HttpStatusCode.NotFound) {
+                call.respond(PebbleContent("404.html", mapOf("static_url_path" to staticPath)))
+            }
+            status(HttpStatusCode.MethodNotAllowed) {
+                call.respond(PebbleContent("405.html", mapOf("static_url_path" to staticPath)))
+            }
+            status(HttpStatusCode.InternalServerError) {
+                call.respond(PebbleContent("500.html", mapOf("static_url_path" to staticPath)))
+            }
+        }
+    }
+
+    private fun Route.defaultRoutes(staticPath: String) {
+        static("$staticPath/css") {
+            resources("assets/static/css")
+        }
+        static("$staticPath/img") {
+            resources("assets/static/img")
+        }
+        static("$staticPath/js") {
+            resources("assets/static/js")
+        }
+    }
+
+    private fun Route.sendRoutes(sendPage: SendPage, staticPath: String) {
+        get("/") {
+            val model = sendPage.model + mapOf("static_url_path" to staticPath)
```

Comment:
Would it make sense to declare `mapOf("static_url_path" to staticPath)` as a constant? Seems to be used in a few places and I guess we'll add more routes.

Reply:
I think a compile time constant won't work, because the path is different for each invocation of the webserver (and they need to be of primitive types in kotlin), but I pushed a commit that factors out map into a final variable.